tests/functional/states/test_kubernetes.py

```diff
@@ -0,0 +1,1426 @@
+            {% endfor %}
+        spec:
+          containers:
+          - name: {{ name }}
+            image: {{ image }}
+            ports:
+            - containerPort: 80
+        """
+    ).strip()
+
+    with pytest.helpers.temp_file(f"{sls}.yml.jinja", contents, state_tree):
+        yield f"salt://{sls}.yml.jinja"
+
+
+@pytest.fixture(params=[True])
+def pod(kubernetes, _pod_spec, request):
+    """
+    Fixture providing a pod for testing
+    """
+    name = random_string("pod-", uppercase=False)
+    namespace = "default"
+
+    # Only create the pod if requested
+    if request.param:
+        ret = kubernetes.pod_present(
+            name=name,
+            namespace=namespace,
+            spec=_pod_spec,
+            wait=True,
+        )
+        assert ret.result is True
+    try:
+        yield {"name": name, "namespace": namespace}
+    finally:
+        ret = kubernetes.pod_absent(name=name, namespace=namespace, wait=True)
+        assert ret.result is True
+        assert ret.comment in ["The pod does not exist", "Pod deleted"]
+
+
+@pytest.mark.parametrize("pod", [False], indirect=True)
+def test_pod_present(kubernetes, pod, _pod_spec):
+    """
+    Test kubernetes.pod_present creates a pod
+    """
+    ret = kubernetes.pod_present(
+        name=pod["name"],
+        namespace=pod["namespace"],
+        spec=_pod_spec,
+        wait=True,
+    )
+
+    assert ret.result is True
+    assert ret.changes
+
+    # Comment out idempotent test for now
+    # TODO: The state module needs fixed to handle proper present functionality
+
+
+# def test_pod_present_idempotency(kubernetes, pod, _pod_spec):
+#     """
+#     Test kubernetes.pod_present is idempotent
+#     """
+#     ret = kubernetes.pod_present(
+#         name=pod["name"],
+#         namespace=pod["namespace"],
+#         spec=_pod_spec,
+#         wait=True,
+#     )
+
+#     assert ret.result is True
+#     assert "already exists" in ret.comment
+#     assert not ret.changes
+
+
+def test_pod_present_test_mode(kubernetes, pod, _pod_spec):
+    """
+    Test kubernetes.pod_present in test mode
+    """
+    ret = kubernetes.pod_present(
+        name=pod["name"],
+        namespace=pod["namespace"],
+        spec=_pod_spec,
+        test=True,
+    )
+
+    assert ret.result is None
+    assert not ret.changes
+
+
+@pytest.mark.parametrize("pod", [False], indirect=True)
+def test_pod_present_template_context(kubernetes, pod, pod_template):
+    """
+    Test kubernetes.pod_present with template_context
+    """
+    template_context = {
+        "name": pod["name"],
+        "namespace": pod["namespace"],
+        "image": "nginx:latest",
+        "labels": {"app": "test"},
+    }
+
+    ret = kubernetes.pod_present(
+        name=pod["name"],
+        namespace=pod["namespace"],
+        source=pod_template,
+        template="jinja",
+        template_context=template_context,
+        wait=True,
+    )
+
+    assert ret.result is True
+    assert ret.changes
```

Comment:
suggestion: As noted above, I think this test should verify that the template has been applied as expected.